odoo/addons/base/tests/test_views.py

```diff
@@ -2213,6 +2213,47 @@ def test_modifier_attribute_using_context(self):
             self.assertEqual(modifiers.get('readonly'), expected.get('readonly'))
             self.assertEqual(modifiers.get('required'), expected.get('required'))
 
+    @common.users('demo')
+    def test_form_attribute_priority(self):
+        """" Specifying the value of an attribute on the form node
+        should take precedence over ACL checks. """
+
+        regular_view = self.View.create({
+            'name': 'foo',
+            'model': 'res.users',
+            'arch': """
+                <form>
+                    <field name="name"/>
+                </form>
+            """,
+        })
+
+        arch = self.env['res.users'].get_view(regular_view.id)['arch']
+        regular_form_node = etree.fromstring(arch).xpath('//form')[0]
+        # demo user does not have any rights on the res.users model
+        # flags should therefore be set to false for create, edit and delete actions
+        self.assertEqual(regular_form_node.get('create'), 'false')
+        self.assertEqual(regular_form_node.get('edit'), 'false')
+        self.assertEqual(regular_form_node.get('delete'), 'false')
+
+        custom_view = self.View.create({
+            'name': 'foo',
+            'model': 'res.users',
+            'arch': """
+                <form edit="1">
+                    <field name="name"/>
+                </form>
+            """,
+        })
+
+        arch = self.env['res.users'].get_view(custom_view.id)['arch']
+        custom_form_node = etree.fromstring(arch).xpath('//form')[0]
+        # same as the regular case, except that we forced edit="1" in the arch
+        # this should take precedence over the ACL in the result of "get_view"
+        self.assertEqual(custom_form_node.get('create'), 'false')
+        self.assertEqual(custom_form_node.get('edit'), '1')
+        self.assertEqual(custom_form_node.get('delete'), 'false')
+
```

Comment:
This unit test doesn't really test what your code change is fixing.

This unit test tests the fact `edit="1"` on the view node as the priority on the ACLs. But this was already working before your fix, this is not what your code change is fixing.

What should should test is that loading the User Preferences form view as the demo user allows the demo user to edit the fields. Given the fact `res.users` has a tricky behavior regarding the edition of its fields despite the ACLs normally prevent it.

So, to me, the unit test you should add should be placed in `odoo/addons/base/tests/test_res_users.py`, as it tests a behavior specific to `res.users` only.
In the unit test, you should load the view `view_users_form_simple_modif` through the `Form` test util, with the demo user as the demo user, then ensure the form is editable, for instance by checking changing fields, such as the user email or timezone, is allowed.

e.g. (untested code)
```py
demo_user = self.env.ref('base.user_demo')
with Form(demo_user.with_user(demo_user), view='base.view_users_form_simple_modif') as form:
    form.email = 'foo@bar.com'
self.assertEqual(demo_user.email, 'foo@bar.com')
```
And check that reverting your fix temporary, removing that `edit="1"` in that `view_users_form_simple_modif`, fails the above test.

Reply:
Hello @beledouxdenis 

I tried a couple things today.

- The test that you provided.

Unfortunately, as the issue lies with the web framework that sets the form view as readonly, this test does not fail without my fix.

- A tour

After discussing this with TDE, we concluded that the best way to test it was a tour.
So I wrote a tour, in the mail module, as base does not have the tour assets.

Tested in local, worked great, life was good.

However, it's mandatory to set your tour tests as post_install.
(See https://github.com/odoo/odoo/blob/16.0/odoo/tests/common.py#L1822 ).
(We asked XDO, it has something to do with the database needing being fully initialized with all menus, otherwise it makes the tours impossible to debug, conclusion is we don't support it).

Making the tour post_install means the tour useless.
As "hr" module is installed, and it completely changes the flow of the user preferences (see my previous comment about that).

So I'm a bit lost here, I don't think I can write a test that will correctly cover this fix.

Unless you have an idea, maybe we could proceed without a test exceptionally?

Thanks again for the guidance.

Reply:
Hello @beledouxdenis 

Actually after discussing this one more time with XDO, we came up with a different approach.
We keep the test in mail and adapt the steps in the hr module so that both flows can be tested.

At first I thought that it would mean we never really test the flow with only the mail module, but that's actually not true as single module builds are done during nightly builds on the runbot.

It also has the advantage of testing this flow when hr is installed as well, which seems like a good thing.